keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+future versions of the api resources introduced in this or other future KEPs.
+For the time being, the AdminNetworkPolicy resource introduced by this KEP will
+never affect north-south traffic, and thus also don't override or bypass NetworkPolicies
+with ipBlock rules that select external traffic.
+
+## Proposal
+
+In order to achieve the three primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `policy.networking.k8s.io` API group:
+- AdminNetworkPolicy
+
+### AdminNetworkPolicy resource
+
+An AdminNetworkPolicy (ANP) resource will help the administrators:
+1. Set strict security rules for the cluster, i.e. a developer CANNOT override
+these rules by creating NetworkPolicies that applies to the same workloads as the
+AdminNetworkPolicy does.
+2. Set baseline security rules that describes default connectivity for cluster
+workloads, which CAN be overridden by developer NetworkPolicies if needed.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, AdminNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. AdminNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the AdminNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all further rules from all 
+  ANPs and instead be enforced by the K8s NetworkPolicies. If there is no K8s 
+  NetworkPolicy rule match, traffic will be allowed.
+- Deny: Traffic that matches a `Deny` rule will be dropped. 
+- Allow: Traffic that matches an `Allow` rule will be allowed.
+
+AdminNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic with malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the `Deny` rules are deleted
+or modified by the admin.
+
+On the other hand, the `Allow` rules can be used to call out traffic in the cluster
+that needs to be allowed for certain components to work as expected (egress to
+CoreDNS for example). Those traffic should not be blocked when developers apply
+NetworkPolicy to their Namespaces which isolates the workloads.
+
+AdminNetworkPolicy `Pass` rules allows an admin to delegate security posture for
+certain traffic to the Namespace owners by overriding any lower priority Allow or 
+Deny rules. For example, allowing intra-tenant traffic can be delegated to tenant 
+admins explicitly by the cluster admin with the use of `Pass` rules.
+
+#### Priority
+
+The policy instances will be ordered based on the numeric priority assigned to the
+ANP. A larger number corresponds to a higher precedence. For example, ANP with `priority`
+set to "1" will be considered as the policy with the lowest precedence, while "999" will
+be considered as highest precedence. Any positive priority, however, will have
```

Comment:
Agreed!

Reply:
> For future-safety, clients may assume that higher values will eventually be allowed, and simply treat it as an int32.

That would be a somewhat-breaking change though, because it means that in a future release, your previously-un-overridable `Priority: 1000` policy might suddenly be outranked.

(I had suggested before that we should allow any int32 value, but just say that the plugin is not required to allow there to be more than 1000 distinct priority values. I guess that would be really awkward though if you actually had to squash the larger priority values into a smaller range...)

Reply:
What if we said that the numeric range was large (e.g. 1M) but there's a practical (and implementation-defined?) limit on the number of unique values a given flow can specify.  Implementations are responsible for mapping the 1M range into their implementation range.  The remapping MAY assign multiple input (ANP) values to the same output (impl) value, as long as those can't be for the same flow.

I feel like even at 1000, impls which support less than 1000 have to do this.  As  long as impl-space < abstract space, a mapping function must exist.

So concretely one rule at prio 500 with a target of "namespaces where tenant=coke" and another rule at prio 600 with a target of "namespaces where tenant=pepsi" are provably disjoint, so both might map to prio 50 in 0-99 implementation.

Reply:
> That would be a somewhat-breaking change though, because it means that in a future release, your previously-un-overridable Priority: 1000 policy might suddenly be outranked.

Good point.  Once we go GA (maybe eve just beta) we hve to lock it in.